src/core/socket-bind.c

```diff
@@ -0,0 +1,205 @@
+
+        if (r == 0) {
+                log_debug_errno(SYNTHETIC_ERRNO(EOPNOTSUPP),
+                                "Not running with unified cgroup hierarchy, BPF is not supported");
+                return 0;
+        }
+
+        if (!bpf_probe_prog_type(BPF_PROG_TYPE_CGROUP_SOCK_ADDR, /*ifindex=*/0)) {
+                log_debug_errno(SYNTHETIC_ERRNO(EOPNOTSUPP),
+                                "BPF program type cgroup_sock_addr is not supported");
+                return 0;
+        }
+
+        r = prepare_socket_bind_bpf(/*unit=*/NULL, /*allow_rules=*/NULL, /*deny_rules=*/NULL, &obj);
+        if (r < 0) {
+                log_debug_errno(r, "BPF based socket_bind is not supported: %m");
+                return 0;
+        }
+
+        return can_link_bpf_program(obj->progs.socket_bind_v4);
+}
+
+int socket_bind_install(Unit *u) {
+        _cleanup_(bpf_link_freep) struct bpf_link *ipv4 = NULL, *ipv6 = NULL;
+        _cleanup_(socket_bind_bpf_freep) struct socket_bind_bpf *obj = NULL;
+        _cleanup_free_ char *cgroup_path = NULL;
+        _cleanup_close_ int cgroup_fd = -1;
+        CGroupContext *cc;
+        int r;
+
+        cc = unit_get_cgroup_context(u);
+        if (!cc)
+                return 0;
+
+        r = cg_get_path(SYSTEMD_CGROUP_CONTROLLER, u->cgroup_path, NULL, &cgroup_path);
+        if (r < 0)
+                return log_unit_error_errno(u, r, "Failed to get cgroup path: %m");
+
+        if (!cc->socket_bind_allow && !cc->socket_bind_deny)
+                return 0;
+
+        r = prepare_socket_bind_bpf(u, cc->socket_bind_allow, cc->socket_bind_deny, &obj);
+        if (r < 0)
+                return log_unit_error_errno(u, r, "Failed to load BPF object: %m");
+
+        cgroup_fd = open(cgroup_path, O_RDONLY | O_CLOEXEC, 0);
+        if (cgroup_fd < 0)
+                return log_unit_error_errno(
+                                u, errno, "Failed to open cgroup=%s for reading", cgroup_path);
+
+        ipv4 = bpf_program__attach_cgroup(obj->progs.socket_bind_v4, cgroup_fd);
+        r = libbpf_get_error(ipv4);
+        if (r != 0)
+                return log_unit_error_errno(u, r, "Failed to link '%s' cgroup-bpf program",
+                                bpf_program__name(obj->progs.socket_bind_v4));
+
+        ipv6 = bpf_program__attach_cgroup(obj->progs.socket_bind_v6, cgroup_fd);
+        r = libbpf_get_error(ipv6);
+        if (r != 0)
+                return log_unit_error_errno(u, r, "Failed to link '%s' cgroup-bpf program",
+                                bpf_program__name(obj->progs.socket_bind_v6));
+
+        if (!u->socket_bind) {
+                u->socket_bind = new(SocketBind, 1);
+                if (!u->socket_bind)
+                        return log_oom();
+                *u->socket_bind = (SocketBind) {};
+        }
+
+        u->socket_bind->ipv4_link = TAKE_PTR(ipv4);
+        u->socket_bind->ipv6_link = TAKE_PTR(ipv6);
```

Comment:
Either new0 should be used, and the explicit initialization dropped,
or the explicit intiialization should be dropped, and the two assignments replaced by structured initialization.
The latter is preferred.
```C
       if (!u->socket_bind) {
                u->socket_bind = new(SocketBind, 1);
                if (!u->socket_bind)
                        return log_oom();
        }

        u->socket_bind = {
                .ipv4_link = TAKE_PTR(ipv4),
                .ipv6_link = TAKE_PTR(ipv6),
        };
```
(Though SocketBind should go away... I'm commenting in case some parts of this code remain.)